### ISSUE TITLE
Megatron Energon Data Loading Pipeline sample randomness; I/O patterns and questions

### ISSUE THREAD

**Issue by bernardhan33** [NVIDIA/Megatron-Energon#110]
Hi Team and @philipp-fischer  ,

This is relating to comment https://github.com/NVIDIA/Megatron-Energon/issues/81#issuecomment-2693297439 that I wanted to gain some deep understandings into how Megatron Energon data loaders shuffles samples with shard ranges, etc. 

Therefore, I prepared the following illustration:

![Image](https://github.com/user-attachments/assets/80992951-93e9-4fe6-a893-d6cea0efa7cb)

And here's my brief understanding of the access patterns:

1. Using the “num_workers”, “max_samples_per_sequence” and the data parallel group size (and potentially more info!), the Megatron Energon data loader determines the sharding allocation to each worker. Each worker will be assigned with a number of mini-shards, where each mini-shard contains adjacent samples from a specific tar file.
2. At the beginning of data loading, each worker randomly selects the “parallel_shard_iters” number of shards to open the iterators. Each iterator will be kept open until the iterator is exhausted.
3. Each shard iterator contributes to serve samples to fill the in-memory buffer of size “shuffle_buffer_size”. The samples in the buffer are shuffled and served. Once a sample is served and dropped from the buffer, the open iterators are randomly selected (with weights) to fetch the next sample and fill the buffer.
4. Each worker process participates in the round-robin fashion of obtaining data samples, just like how a native PyTorch dataloader works.

And here are my questions:
1. Could anyone confirm if my illustration and understanding above are accurate? Please correct me if I missed anything.
2. Recently we are running a workload with the [recommended settings](https://nvidia.github.io/Megatron-Energon/basic/quickstart.html) of `shuffle_buffer_size=100` and `max_samples_per_sequence=100` with an image dataset where each tar file contains more than 3000 images.  We are prototyping our workload on a single node that essentially has the data parallel group of 4 ranks. Therefore, it is natural for the following sharding allocation for rank 0 worker 0:

```
rank=0, worker=0: shard_range=[pretrain-0.tar[0, 98), pretrain-0.tar[98, 197), pretrain-0.tar[197, 295), ...<28>, pretrain-1.tar[302, 403), pretrain-1.tar[403, 504), pretrain-1.tar[504, 605)] sum(count)=3367
```

Then we noticed that because of the recent addition of the LRU cache, the file descriptor related with `pretrain-0.tar` is kept open but due to the way that the buffer is refilled (No. 3 understanding above), there is a file seek for every sample read. For example, assuming that the first 3 shard ranges are selected, the same file descriptor can be used to fetch the 0th sample (shard range 1) --> 197th sample (shard range 3) --> 98th sample (shard range 2) --> 1st sample (shard range 1) and onwards. Is this a correct understanding?

If so, I would think that I can just go ahead to set `max_samples_per_sequence=1` to get a better sample randomness while not sacrificing the read performance since it has already been mostly random reads within the file based on the above understanding. And by doing that, it doesn't really matter if I keep `shuffle_buffer_size=100` or just `shuffle_buffer_size=1` because I would get the similar amount of randomness. Did I miss anything here?

Thanks a ton!

**Comment by voegtlel** [comment]
Hi Bernard,

thanks for your report.

First, regarding your understanding of our access pattern: You're quite correct, this is what's happening in Energon. There is only one detail missing, which is not related to your issue though: The slices are distributed in a [binary interleaved pattern](https://github.com/NVIDIA/Megatron-Energon/blob/develop/src/megatron/energon/flavors/webdataset/sharder.py#L140) to distribute the slices better across ranks (which may have multiple workers) and have a [rotation offset if blending multiple datasets](https://github.com/NVIDIA/Megatron-Energon/blob/develop/src/megatron/energon/flavors/webdataset/sharder.py#L198) to balance out small blended datasets across all ranks

Also, we agree with your assessment regarding the seeks. We discussed if opening a file per sequence or sharing the file handle is better. Although, for a typical filesystem, os caching should help here, i.e. that the consecutive read after a seek could already be paged. We actually just benchmarked the reading speed for this, and saw that the page cache has a positive effect.

To give you a heads up: We recently ran some benchmarks and found that even fully random access pattern with disabled page cache can work without much impact. We are continuing to investigate that option. Would you be interested to talk in a call or so?

In general, I see 1/3 more "long seeks" (more than +8KiB forward or going -0B backwards, i.e. possible page fails) if we keep the current file handle cache, compared to changing the cache to consider sequences on a quite realistic dataset (tested with 500k samples and 5M samples). For us, the IO time is approximately the same, likely because of os page caching (also considering that decoding images outweights the reading time strongly).

BR
@voegtlel & @philipp-fischer

**Comment by bernardhan33** [comment]
Thanks @voegtlel for the detailed response!

First, thank you for confirming my understanding. I probably missed the blending part as my workload used a single dataset source. But when you mentioned _"The slices are distributed in a [binary interleaved pattern](https://github.com/NVIDIA/Megatron-Energon/blob/develop/src/megatron/energon/flavors/webdataset/sharder.py#L140) to distribute the slices better across ranks (which may have multiple workers)"_, I noticed that based on my logs, the same shard ranges are always allocated to the same rank and the same worker though, such as
```
rank=0, worker=0: shard_range=[pretrain-0.tar[0, 98), pretrain-0.tar[98, 197), pretrain-0.tar[197, 295), ...<28>, pretrain-1.tar[302, 403), pretrain-1.tar[403, 504), pretrain-1.tar[504, 605)] sum(count)=3367
```
Did I miss understand anything here?

Secondly, agreed that "a file handle per sequence" or "a file handle per file" is an interesting design choice to make (which @philipp-fischer and I had conversations over in https://github.com/NVIDIA/Megatron-Energon/issues/81). And thanks for reminding me that the page cache can be helpful for the current "a file handle per file" behavior! To share more info -- My workloads run on a typical Debian OS with 128KB kernel readahead, with dataset stored in remote object store (GCS) and a FUSE adapter ([GCSFUSE](https://cloud.google.com/storage/docs/cloud-storage-fuse/overview)). My dataset also contains large images (1MB). So with all these, somehow the kernel is not requesting "ReadFile" beyond the end offset of the sample. But we can always configure the kernel readahead to test.

I'm happy to schedule a call to connect, share more findings and hear back from you. Could you email me at bernardhan@google.com and we can schedule some time?

Lastly, I would love to hear more on how your team creates a large dataset that consists, for example, 128M samples. I'm in the process of creating one as well to test the performance and I'm currently generating 128M individual images first, converting them into WDS and preparing to use "energon prepare ." as [NeMo recommends](https://docs.nvidia.com/nemo-framework/user-guide/latest/vlms/energondataprep.html). And I'm looking into using 1MB images so 128M of them gives 128 TB. I can generate the images and convert them into WDS on the fly but it seems to me that the only way of converting WDS into Megatron-Energon format is through `energon prepare .` which was implemented by [prepare.py](https://github.com/NVIDIA/Megatron-Energon/blob/develop/src/megatron/energon/tools/prepare.py). I wonder if there are some utility functions that would allow me to generate the Megatron-Energon compatiable dataset on the fly (in a "streaming" fashion)?

Thanks a ton!

**Comment by philipp-fischer** [comment]
### Meeting
You should receive an invite from Yu Ding.

### Binary interleaved pattern
I'm sorry, there was a mistake about how the binary interleaved pattern is used. We use this pattern to distribute the "imbalance" caused by the number of samples not being divisible by the number of global workers. So the number of samples assigned to each worker is determined by the pattern. However, we still assign consecutive sequences of samples to workers for the balanced part to make use of OS read-ahead and caching. Here's an example to explain how this works. Each column is a worker (worker number in white header row). We have seven workers, one blue dataset with 9 samples and one green dataset with 10 samples.

![Image](https://github.com/user-attachments/assets/265f43b5-abe4-4d21-a825-c55c61c60ee1)

### GCS Access
I see that you are using GCS and mounting that using GCSFUSE. Energon relies on the multi-storage-client MSC which support GCS natively, so you might want to consider using that directly without FUSE.
https://nvidia.github.io/multi-storage-client/config/index.html#gcs
Instead of passing paths like `/my/data/dataset`, you would then pass a URL like `msc://mygcsstore/path/to/dataset` to `get_train_dataset` or even inside your blended metadatasets.

### Energon prepare
The `energon prepare` command is an interactive convenience tool, which does several things:
- It scans your folder for all tar files and creates an index for each (`.tar.idx`). It also creates an `sqlite` index which is used in newer versions of energon. A `.info.yaml` file will also be created which contains the samples counts per tar.
- The tool also walks you through creating a `dataset.yaml` file and a sample loader.

In your case, you would want to automate the first stage on many datasets. The second stage, you can customize as you wish and create those files programmatically yourself.
Note that we recommend few large shards over many small ones. So please try not to have hundreds of thousands of shards, but rather maximum 10k.

To automate the first part, please directly use `BaseWebdatasetFactory.prepare_dataset` in a similar way as used [here](https://github.com/NVIDIA/Megatron-Energon/blob/feature/samples_on_demand/src/megatron/energon/tools/prepare.py#L210)
